Sources/PackageRegistry/RegistryManager.swift

```diff
@@ -0,0 +1,293 @@
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+ */
+
+import Basics
+import PackageLoading
+import PackageModel
+
+import TSCBasic
+import TSCUtility
+
+import struct Foundation.URL
+import struct Foundation.URLComponents
+import struct Foundation.URLQueryItem
+
+import Dispatch
+
+public enum RegistryError: Error {
+    case invalidOperation
+    case invalidResponse
+    case invalidURL
+    case invalidChecksum(expected: String, actual: String)
+}
+
+fileprivate let userAgent = "SwiftPackageManager/\(ToolsVersion.currentToolsVersion)"
+
+public final class RegistryManager {
+    internal static var clientFactory: () -> HTTPClientProtocol = {
+        var configuration = HTTPClientConfiguration()
+        configuration.followRedirects = false
+
+        return HTTPClient(configuration: configuration, handler: nil, diagnosticsEngine: nil)
+    }
+
+    let registryBaseURL: Foundation.URL
+
+    init(registryBaseURL: Foundation.URL) {
+        self.registryBaseURL = registryBaseURL
+    }
+
+    public class func discover(
+        for package: PackageReference,
+        completion: @escaping (Result<RegistryManager, Error>) -> Void
+    ) {
+        guard let url = URL(string: "https://\(package.identity)") else {
+            return completion(.failure(RegistryError.invalidURL))
+        }
+
+        let request = HTTPClient.Request(
+            method: .head,
+            url: url,
+            headers: [
+                "Accept": "application/vnd.swift.registry.v1",
+                "User-Agent": userAgent
+            ]
+        )
+
+        let client = clientFactory()
+        client.execute(request) { result in
+            switch result {
+            case .success(let response):
+                if response.statusCode == 303,
+                   response.headers.get("Content-Version").first == "1",
+                   let location = response.headers.get("Location").first,
+                   let redirectedURL = URL(string: location, relativeTo: url)
+                {
+                    let manager = RegistryManager(registryBaseURL: redirectedURL)
+                    completion(.success(manager))
+                } else {
+                    completion(.failure(RegistryError.invalidResponse))
+                }
+            case .failure(let error):
+                completion(.failure(error))
+            }
+        }
+    }
+
+    public func fetchVersions(of package: PackageReference, completion: @escaping (Result<[Version], Error>) -> Void) {
+        let request = HTTPClient.Request(
+            method: .get,
+            url: registryBaseURL,
+            headers: [
+                "Accept": "application/vnd.swift.registry.v1+json",
+                "User-Agent": userAgent
+            ]
+        )
+
+        let client = Self.clientFactory()
+        client.execute(request) { result in
+            switch result {
```

Comment:
in some cases you can use result.tryMap instead of switching for the failure case